rdfproxy/utils/utils.py

```diff
@@ -1,90 +1,76 @@
+def _list_p(obj: type | None) -> bool:
+    """Check if obj is a list type."""
+    return _type_p(obj, list)
 
-    Example:
 
-        class SimpleModel(BaseModel):
-            x: int
-            y: int
+def _list_basemodel_p(obj: type | None) -> bool:
+    """Check if a type is list[pydantic.BaseModel]."""
+    return (get_origin(obj) is list) and all(
+        issubclass(cls, BaseModel) for cls in get_args(obj)
+    )
 
 
-        class NestedModel(BaseModel):
-            a: str
-            b: SimpleModel
+def _collect_values_from_bindings(
+    binding_name: str,
+    bindings: Iterable[dict],
+    predicate: Callable[[Any], bool] = lambda x: x is not None,
+) -> list:
+    """Scan bindings for a key binding_name and collect unique predicate-compliant values.
 
+    Note that element order is important for testing, so a set cast won't do.
+    """
+    values = dict.fromkeys(
+        value
+        for binding in bindings
+        if predicate(value := binding.get(binding_name, None))
+    )
+    return list(values)
 
-        class ComplexModel(BaseModel):
-            p: str
-            q: NestedModel
 
+def _get_key_from_metadata(v: FieldInfo) -> str | None:
+    """Try to get a SPARQLBinding object from a field's metadata attribute.
 
-    model = instantiate_model_from_kwargs(ComplexModel, x=1, y=2, a="a value", p="p value")
-    print(model)  # p='p value' q=NestedModel(a='a value', b=SimpleModel(x=1, y=2))
+    Helper for _generate_binding_pairs.
     """
-
-    def _get_key_from_metadata(v: FieldInfo):
-        """Try to get a SPARQLBinding object from a field's metadata attribute.
-
-        Helper for _generate_binding_pairs.
-        """
-        try:
-            value = next(filter(lambda x: isinstance(x, SPARQLBinding), v.metadata))
-            return value
-        except StopIteration:
-            return None
-
-    def _generate_binding_pairs(
-        model: type[_TModelInstance], **kwargs
-    ) -> Iterator[tuple[str, Any]]:
-        """Get the bindings needed for model instantation.
-
-        The function traverses model.model_fields
-        and constructs binding pairs by either getting values from kwargs or field defaults.
-        For model fields the recursive clause runs.
-        """
-        for k, v in model.model_fields.items():
-            if isinstance(v.annotation, type(BaseModel)):
-                value = v.annotation(
-                    **dict(_generate_binding_pairs(v.annotation, **kwargs))
-                )
-            else:
-                binding_key = _get_key_from_metadata(v) or k
-                value = kwargs.get(binding_key, v.default)
-
-            yield k, value
-
-    bindings = dict(_generate_binding_pairs(model, **kwargs))
-    return model(**bindings)
+    try:
+        value = next(filter(lambda x: isinstance(x, SPARQLBinding), v.metadata))
+        return value
+    except StopIteration:
+        return None
```

Comment:
This could be rewritten as `return next(filter(lambda x: isinstance(x, SPARQLBinding), v.metadata), None)`
(or, even simpler, use `next(filter(lambda x: isinstance(x, SPARQLBinding), v.metadata), k)` in the two lines where this function is used and drop the whole function)

Reply:
So true, I totally forgot about `next`'s default argument! Will implement right away!

Reply:
`_get_key_from_metadata` takes a default argument now which is passed to `next`.